File: working/recursividad/backtracking/sudoku.py
class Sudoku:
    Sudoku.N = 9


    def __init__(self,posicionesFijas):
        self.__posicionesFijas = posicionesFijas

        self.__estado = []
        for item in self.__posicionesFijas:
            self.__estado.append(item)

        self.__posicionActual = 0
        self.__movimientos = 0
        self.__noSoluciones = 0


    def __str__(self):
        result = "Tablero:\n"
        for i in range(Sudoku.N*Sudoku.N):
            if i % (Sudoku.N*(Sudoku.N//3)) == 0:
                for j in range(Sudoku.N):
                    if j % (Sudoku.N//3) == 0:
                        result += '-+'
                    result += '----'
                result += "-+\n"
            if i % (Sudoku.N//3) == 0:
                result += ' |'
            if self.__estado['posicionesFijas'][i] != 0:
                result += f' *{self.__estado["tablero"][i]:2d}'
            else:
                result += f'  {self.__estado["tablero"][i]:2d}'
            if i % Sudoku.N == Sudoku.N-1:
                result += " |\n"

        for j in range(Sudoku.N):
            if j % (Sudoku.N//3) == 0:
                result += '-+'
            result += '----'

        result += "-+\n"
        result += f'Movimientos: self.__movimientos'
        result += 'Soluciones: ', self.__noSoluciones'
        return result


    def esUnaSolucion(self):
        return self.__posicionActual >= Sudoku.N*Sudoku.N


    def construirCandidatos(self):
        candidatos = []
        if self.__posicionesFijas[self.__posicionActual] != 0:
            candidatos.append(self.__posicionesFijas[self.__posicionActual])
        else:
            for i in range(1,Sudoku.N+1):
                if self.esValido(i):
                    candidatos.append(i)
        return candidatos


    def getFila(self,posicion,anchuraCuadrado):
        return posicion // anchuraCuadrado


    def getColumna(self,posicion,anchuraCuadrado):
        return posicion % anchuraCuadrado


    def getPosicion(self,fila,columna,anchuraCuadrado):
        return fila * anchuraCuadrado + columna


def esValido( self.__estado, num ):
    esta = False

    # misma fila
    k = 0
    while not esta and k < Sudoku.N:
        esta = self.__estado[getPosicion( getFila(self.__estado['posicionActual'],Sudoku.N),k,Sudoku.N )] == num
        k += 1

    # misma columna
    k = 0
    while not esta and k < Sudoku.N:
        esta = self.__estado[getPosicion( k, getColumna(self.__estado['posicionActual'],Sudoku.N),Sudoku.N )] == num
        k += 1

    # mismo cuadrante
    k = 0
    refFila = getFila(self.__estado['posicionActual'],Sudoku.N) // (Sudoku.N // 3) * 3
    refColumna = getColumna(self.__estado['posicionActual'],Sudoku.N) // (Sudoku.N // 3) * 3
    while not esta and k < Sudoku.N:
        esta = self.__estado[getPosicion( refFila + getFila( k, Sudoku.N // 3 ), refColumna + getColumna( k, Sudoku.N // 3 ), Sudoku.N )] == num
        k += 1

    return not esta


def extender( self.__estado, candidato ):
    self.__estado[self.__estado['posicionActual']] = candidato
    self.__estado['posicionActual'] += 1


def volver( self.__estado ):
    self.__estado['posicionActual'] -= 1
    if self.__estado['posicionesFijas'][self.__estado['posicionActual']] == 0:
        self.__estado[self.__estado['posicionActual']] = 0


def backtrack( self.__estado ):
    self.__estado['movimientos'] += 1
    if esUnaSolucion( self.__estado ):
        self.__estado['noSoluciones'] += 1
        procesarSolucion( self.__estado )
    else:
        candidatos = construirCandidatos( self.__estado )
        for candidato in candidatos:
            extender( self.__estado, candidato )
            backtrack( self.__estado );
            volver( self.__estado );

self.__estado = inicializa()
backtrack(self.__estado)

File: working/recursividad/backtracking/sudoku.rb
class Sudoku
    N = 9


    def initialize(posicionesFijas)
        @posicionesFijas = posicionesFijas

        @estado = []
        for item in @posicionesFijas
            @estado << item
        end

        @posicionActual = 0
        @movimientos = 0
        @soluciones = 0
    end

    
    def to_s
        result = ""
        result += "Tablero:\n"
        (N*N).times do |i|
            if i % (N*(N/3)) == 0
                N.times do |j|
                    if j % (N/3) == 0
                        result += '-+'
                    end
                    result += '---'
                end
                result += "-+\n"
            end
            if i % (N/3) == 0
                result += ' |'
            end
            if @posicionesFijas[i] != 0
                result += " *#{@estado[i]}"
            else
                result += "  #{@estado[i]}"
            end
            if i % N == N-1
                result += " |\n"
            end
        end

        N.times do |j|
            if j % (N/3) == 0
                result += '-+'
            end
            result += '---'
        end

        result += "-+\n"
        result += "Movimientos: #{@movimientos} / Soluciones: #{@soluciones}"
        return result
    end

    def esUnaSolucion?
        return @posicionActual >= N*N
    end

    def construirCandidatos
        candidatos = []
        if @posicionesFijas[@posicionActual] != 0
            candidatos << @posicionesFijas[@posicionActual]
        else
            (1..N).each do |i|
                if esValido?(i)
                    candidatos << i
                end
            end
        end
        return candidatos
    end

    def extender( candidato )
        @estado[@posicionActual] = candidato
        @posicionActual += 1
    end

    def volver
        @posicionActual -= 1
        if @posicionesFijas[@posicionActual] == 0
            @estado[@posicionActual] = 0
        end
    end

    def getFila( posicion, anchuraCuadrado )
        return posicion / anchuraCuadrado
    end

    def getColumna( posicion, anchuraCuadrado )
        return posicion % anchuraCuadrado
    end

    def getPosicion( fila, columna, anchuraCuadrado )
        return fila * anchuraCuadrado + columna
    end

    def esValido?( num )
        esta = false

        # misma fila
        k = 0
        while !esta and k < N
            esta = @estado[getPosicion( getFila(@posicionActual,N),k,N )] == num
            k += 1
        end

        # misma columna
        k = 0
        while !esta and k < N
            esta = @estado[getPosicion( k, getColumna(@posicionActual,N),N )] == num
            k += 1
        end

        # mismo cuadrante
        k = 0
        refFila = getFila(@posicionActual,N) / (N / 3) * 3
        refColumna = getColumna(@posicionActual,N) / (N / 3) * 3
        while !esta and k < N
            esta = @estado[getPosicion( refFila + getFila( k, N / 3 ), refColumna + getColumna( k, N / 3 ), N )] == num
            k += 1
        end
        return !esta
    end

    def backtrack
        @movimientos += 1
        if esUnaSolucion?
            @soluciones += 1
            puts self
        else
            candidatos = construirCandidatos
            candidatos.each do |candidato|
                extender( candidato )
                backtrack
                volver
            end
        end
    end

    def backtrack2
        @movimientos += 1
        candidatos = construirCandidatos
        candidatos.each do |candidato|
            extender( candidato )
            if esUnaSolucion?
                @soluciones += 1
                puts self
            elsif @posicionActual < N*N
                backtrack2
            end
            volver
        end
    end
end


s = Sudoku.new(
    [
        0,6,0,7,0,8,1,9,2,
        1,0,5,2,0,0,0,0,7,
        0,2,0,0,0,6,0,0,0,
        0,5,0,9,3,0,0,4,0,
        0,0,6,5,0,2,7,8,0,
        9,7,0,0,0,0,3,2,5,
        0,0,7,4,0,0,8,0,6,
        8,9,4,0,7,0,0,0,0,
        0,1,0,3,0,0,0,7,4,
    ]
)
s.backtrack



=begin
[
    5,3,0,0,0,0,0,0,0,
    6,0,0,0,0,5,0,0,0,
    0,9,8,0,0,0,0,6,0,
    8,0,0,0,6,0,0,0,3,
    4,0,0,8,0,3,0,0,1,
    7,0,0,0,2,0,0,0,6,
    0,6,0,0,0,0,2,8,0,
    0,0,0,4,1,9,0,0,5,
    0,0,0,0,8,0,0,7,0
]
=end
